File: src/data/courses/lld/real-world-problems/amazon-favorites/design-parking-lot/code/java/ParkingLotSystem.java
import java.time.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.*;
import java.util.concurrent.locks.*;

// ============================================================
// Enumerations
// ============================================================

enum VehicleType {
    MOTORCYCLE, CAR, TRUCK
}

enum SpotType {
    COMPACT, REGULAR, LARGE
}

enum TicketStatus {
    ACTIVE, PAID, LOST
}

enum PaymentMethod {
    CASH, CREDIT_CARD, DIGITAL_WALLET
}

// ============================================================
// Vehicle Class
// ============================================================

class Vehicle {
    private final String licensePlate;
    private final VehicleType vehicleType;
    private final String color;

    public Vehicle(String licensePlate, VehicleType vehicleType, String color) {
        this.licensePlate = licensePlate;
        this.vehicleType = vehicleType;
        this.color = color;
    }

    public VehicleType getVehicleType() { return vehicleType; }
    public String getLicensePlate() { return licensePlate; }
    public String getColor() { return color; }
}

// ============================================================
// ParkingSpot Class
// ============================================================

class ParkingSpot {
    private final String spotId;
    private final SpotType spotType;
    private final int floor;
    private final AtomicBoolean isOccupied;
    private Vehicle vehicle;
    private final ReentrantLock spotLock;  // Thread safety

    public ParkingSpot(String spotId, SpotType spotType, int floor) {
        this.spotId = spotId;
        this.spotType = spotType;
        this.floor = floor;
        this.isOccupied = new AtomicBoolean(false);
        this.vehicle = null;
        this.spotLock = new ReentrantLock();
    }

    // Check if vehicle can fit in this spot
    public boolean canFitVehicle(VehicleType vehicleType) {
        switch (vehicleType) {
            case MOTORCYCLE:
                return true;  // Can park anywhere
            case CAR:
                return spotType == SpotType.REGULAR || spotType == SpotType.LARGE;
            case TRUCK:
                return spotType == SpotType::LARGE;
            default:
                return false;
        }
    }

    // Thread-safe vehicle assignment
    public boolean assignVehicle(Vehicle vehicle) {
        spotLock.lock();
        try {
            if (isOccupied.get()) {
                return false;  // Already occupied
            }
            
            if (!canFitVehicle(vehicle.getVehicleType())) {
                return false;  // Vehicle doesn't fit
            }
            
            this.vehicle = vehicle;
            isOccupied.set(true);
            return true;
        } finally {
            spotLock.unlock();
        }
    }

    // Thread-safe vehicle removal
    public void removeVehicle() {
        spotLock.lock();
        try {
            this.vehicle = null;
            isOccupied.set(false);
        } finally {
            spotLock.unlock();
        }
    }

    public boolean isOccupied() { return isOccupied.get(); }
    public SpotType getSpotType() { return spotType; }
    public String getSpotId() { return spotId; }
    public int getFloor() { return floor; }
}

// ============================================================
// ParkingTicket Class
// ============================================================

class ParkingTicket {
    private final String ticketId;
    private final Vehicle vehicle;
    private final ParkingSpot spot;
    private final LocalDateTime entryTime;
    private LocalDateTime exitTime;
    private TicketStatus status;

    public ParkingTicket(String ticketId, Vehicle vehicle, ParkingSpot spot) {
        this.ticketId = ticketId;
        this.vehicle = vehicle;
        this.spot = spot;
        this.entryTime = LocalDateTime.now();
        this.status = TicketStatus.ACTIVE;
    }

    // Calculate parking duration in hours
    public long calculateDuration() {
        LocalDateTime end = (status == TicketStatus.ACTIVE) ? LocalDateTime.now() : exitTime;
        long hours = Duration.between(entryTime, end).toHours();
        return Math.max(1, hours);  // Minimum 1 hour
    }

    public void markExit() {
        this.exitTime = LocalDateTime.now();
        this.status = TicketStatus.PAID;
    }

    public String getTicketId() { return ticketId; }
    public Vehicle getVehicle() { return vehicle; }
    public ParkingSpot getSpot() { return spot; }
    public TicketStatus getStatus() { return status; }
    public void setStatus(TicketStatus status) { this.status = status; }
}

// ============================================================
// Strategy Pattern: Pricing Strategy
// ============================================================

interface PricingStrategy {
    double calculateFee(VehicleType type, long hours);
}

class HourlyPricingStrategy implements PricingStrategy {
    @Override
    public double calculateFee(VehicleType type, long hours) {
        double hourlyRate;
        switch (type) {
            case MOTORCYCLE: hourlyRate = 2.0; break;
            case CAR: hourlyRate = 4.0; break;
            case TRUCK: hourlyRate = 6.0; break;
            default: hourlyRate = 4.0;
        }
        return hourlyRate * hours;
    }
}

// ============================================================
// Payment Class
// ============================================================

class Payment {
    private final double amount;
    private final PaymentMethod paymentMethod;
    private final LocalDateTime timestamp;
    private boolean processed;

    public Payment(double amount, PaymentMethod paymentMethod) {
        this.amount = amount;
        this.paymentMethod = paymentMethod;
        this.timestamp = LocalDateTime.now();
        this.processed = false;
    }

    public boolean processPayment() {
        // Simulate payment processing
        processed = true;
        System.out.println("Payment of $" + amount + " processed successfully.");
        return true;
    }

    public double getAmount() { return amount; }
    public boolean isProcessed() { return processed; }
}

// ============================================================
// Observer Pattern: DisplayBoard
// ============================================================

class DisplayBoard {
    private final int floorNumber;
    private final Map<SpotType, Integer> availableSpots;

    public DisplayBoard(int floorNumber) {
        this.floorNumber = floorNumber;
        this.availableSpots = new ConcurrentHashMap<>();
        availableSpots.put(SpotType.COMPACT, 0);
        availableSpots.put(SpotType.REGULAR, 0);
        availableSpots.put(SpotType.LARGE, 0);
    }

    public void update(SpotType type, int count) {
        availableSpots.put(type, count);
        display();
    }

    public void display() {
        System.out.println("Floor " + floorNumber + " - Available Spots: " +
                "Compact=" + availableSpots.get(SpotType.COMPACT) + ", " +
                "Regular=" + availableSpots.get(SpotType.REGULAR) + ", " +
                "Large=" + availableSpots.get(SpotType.LARGE));
    }
}

// ============================================================
// ParkingFloor Class
// ============================================================

class ParkingFloor {
    private final int floorNumber;
    private final Map<String, ParkingSpot> spots;
    private final DisplayBoard displayBoard;
    private final ReentrantReadWriteLock floorLock;

    public ParkingFloor(int floorNumber) {
        this.floorNumber = floorNumber;
        this.spots = new ConcurrentHashMap<>();
        this.displayBoard = new DisplayBoard(floorNumber);
        this.floorLock = new ReentrantReadWriteLock();
    }

    public void addSpot(ParkingSpot spot) {
        floorLock.writeLock().lock();
        try {
            spots.put(spot.getSpotId(), spot);
            updateDisplayBoard();
        } finally {
            floorLock.writeLock().unlock();
        }
    }

    // Find available spot for vehicle type
    public ParkingSpot getAvailableSpot(VehicleType vehicleType) {
        floorLock.readLock().lock();
        try {
            for (ParkingSpot spot : spots.values()) {
                if (!spot.isOccupied() && spot.canFitVehicle(vehicleType)) {
                    return spot;
                }
            }
            return null;
        } finally {
            floorLock.readLock().unlock();
        }
    }

    public void updateDisplayBoard() {
        // Count available spots by type
        Map<SpotType, Integer> counts = new EnumMap<>(SpotType.class);
        counts.put(SpotType.COMPACT, 0);
        counts.put(SpotType.REGULAR, 0);
        counts.put(SpotType.LARGE, 0);

        for (ParkingSpot spot : spots.values()) {
            if (!spot.isOccupied()) {
                counts.put(spot.getSpotType(), counts.get(spot.getSpotType()) + 1);
            }
        }

        // Update display board
        for (Map.Entry<SpotType, Integer> entry : counts.entrySet()) {
            displayBoard.update(entry.getKey(), entry.getValue());
        }
    }

    public int getFloorNumber() { return floorNumber; }
}

// ============================================================
// Singleton Pattern: ParkingLot
// ============================================================

class ParkingLot {
    private static volatile ParkingLot instance;
    private static final Object lock = new Object();

    private final String name;
    private final String address;
    private final List<ParkingFloor> floors;
    private final Map<String, ParkingTicket> activeTickets;
    private final PricingStrategy pricingStrategy;
    private final AtomicInteger ticketCounter;
    private final ReentrantLock lotLock;

    // Private constructor for Singleton
    private ParkingLot(String name, String address) {
        this.name = name;
        this.address = address;
        this.floors = new CopyOnWriteArrayList<>();
        this.activeTickets = new ConcurrentHashMap<>();
        this.pricingStrategy = new HourlyPricingStrategy();
        this.ticketCounter = new AtomicInteger(0);
        this.lotLock = new ReentrantLock();
    }

    // Thread-safe Singleton instance (Double-checked locking)
    public static ParkingLot getInstance(String name, String address) {
        if (instance == null) {
            synchronized (lock) {
                if (instance == null) {
                    instance = new ParkingLot(name, address);
                }
            }
        }
        return instance;
    }

    public static ParkingLot getInstance() {
        return getInstance("Main Parking Lot", "123 Main St");
    }

    public void addFloor(ParkingFloor floor) {
        floors.add(floor);
    }

    // Find available spot across all floors
    public ParkingSpot getAvailableSpot(VehicleType vehicleType) {
        for (ParkingFloor floor : floors) {
            ParkingSpot spot = floor.getAvailableSpot(vehicleType);
            if (spot != null) {
                return spot;
            }
        }
        return null;
    }

    // Process vehicle entry
    public ParkingTicket processEntry(Vehicle vehicle) throws Exception {
        lotLock.lock();
        try {
            // Find available spot
            ParkingSpot spot = getAvailableSpot(vehicle.getVehicleType());
            if (spot == null) {
                throw new Exception("No available spot for vehicle type");
            }

            // Assign vehicle to spot
            if (!spot.assignVehicle(vehicle)) {
                throw new Exception("Failed to assign vehicle to spot");
            }

            // Generate ticket
            String ticketId = "T" + ticketCounter.incrementAndGet();
            ParkingTicket ticket = new ParkingTicket(ticketId, vehicle, spot);
            activeTickets.put(ticketId, ticket);

            System.out.println("Vehicle " + vehicle.getLicensePlate() +
                    " parked at spot " + spot.getSpotId() +
                    " on floor " + spot.getFloor());

            // Update display boards
            for (ParkingFloor floor : floors) {
                floor.updateDisplayBoard();
            }

            return ticket;
        } finally {
            lotLock.unlock();
        }
    }

    // Process vehicle exit
    public Payment processExit(ParkingTicket ticket) throws Exception {
        lotLock.lock();
        try {
            if (ticket.getStatus() != TicketStatus.ACTIVE) {
                throw new Exception("Ticket is not active");
            }

            // Calculate fee
            long hours = ticket.calculateDuration();
            double fee = pricingStrategy.calculateFee(ticket.getVehicle().getVehicleType(), hours);

            // Create payment
            Payment payment = new Payment(fee, PaymentMethod.CREDIT_CARD);

            // Process payment
            if (payment.processPayment()) {
                // Mark ticket as paid
                ticket.markExit();

                // Free the spot
                ticket.getSpot().removeVehicle();

                // Remove from active tickets
                activeTickets.remove(ticket.getTicketId());

                System.out.println("Vehicle " + ticket.getVehicle().getLicensePlate() +
                        " exited. Duration: " + hours + " hours, Fee: $" + fee);

                // Update display boards
                for (ParkingFloor floor : floors) {
                    floor.updateDisplayBoard();
                }
            }

            return payment;
        } finally {
            lotLock.unlock();
        }
    }

    public void displayStatus() {
        System.out.println("\n=== Parking Lot Status ===");
        System.out.println("Name: " + name);
        System.out.println("Active Tickets: " + activeTickets.size());
        for (ParkingFloor floor : floors) {
            floor.updateDisplayBoard();
        }
        System.out.println("========================\n");
    }
}

// ============================================================
// Main - Example Usage
// ============================================================

public class ParkingLotSystem {
    public static void main(String[] args) {
        try {
            // Get parking lot instance (Singleton)
            ParkingLot parkingLot = ParkingLot.getInstance("City Center Parking", "Downtown");

            // Create floors
            ParkingFloor floor1 = new ParkingFloor(1);
            ParkingFloor floor2 = new ParkingFloor(2);

            // Add spots to floor 1
            floor1.addSpot(new ParkingSpot("F1-C1", SpotType.COMPACT, 1));
            floor1.addSpot(new ParkingSpot("F1-C2", SpotType.COMPACT, 1));
            floor1.addSpot(new ParkingSpot("F1-R1", SpotType.REGULAR, 1));
            floor1.addSpot(new ParkingSpot("F1-R2", SpotType.REGULAR, 1));
            floor1.addSpot(new ParkingSpot("F1-L1", SpotType.LARGE, 1));

            // Add spots to floor 2
            floor2.addSpot(new ParkingSpot("F2-C1", SpotType.COMPACT, 2));
            floor2.addSpot(new ParkingSpot("F2-R1", SpotType.REGULAR, 2));
            floor2.addSpot(new ParkingSpot("F2-L1", SpotType.LARGE, 2));

            // Add floors to parking lot
            parkingLot.addFloor(floor1);
            parkingLot.addFloor(floor2);

            parkingLot.displayStatus();

            // Create vehicles
            Vehicle car1 = new Vehicle("ABC123", VehicleType.CAR, "Red");
            Vehicle bike1 = new Vehicle("XYZ789", VehicleType.MOTORCYCLE, "Black");
            Vehicle truck1 = new Vehicle("TRK456", VehicleType.TRUCK, "White");

            // Process entries
            ParkingTicket ticket1 = parkingLot.processEntry(car1);
            ParkingTicket ticket2 = parkingLot.processEntry(bike1);
            ParkingTicket ticket3 = parkingLot.processEntry(truck1);

            parkingLot.displayStatus();

            // Simulate some time passing
            System.out.println("\n[Simulating time passage...]\n");
            Thread.sleep(1000);

            // Process exits
            parkingLot.processExit(ticket1);
            parkingLot.processExit(ticket2);

            parkingLot.displayStatus();

            // Try to park another car
            Vehicle car2 = new Vehicle("DEF456", VehicleType.CAR, "Blue");
            ParkingTicket ticket4 = parkingLot.processEntry(car2);

            parkingLot.displayStatus();

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: src/data/courses/lld/real-world-problems/amazon-favorites/design-parking-lot/code/python/parking_lot_system.py
from enum import Enum
from datetime import datetime, timedelta
from typing import Optional, List, Dict
from threading import Lock, RLock
from abc import ABC, abstractmethod
import time

# ============================================================
# Enumerations
# ============================================================

class VehicleType(Enum):
    MOTORCYCLE = "MOTORCYCLE"
    CAR = "CAR"
    TRUCK = "TRUCK"

class SpotType(Enum):
    COMPACT = "COMPACT"
    REGULAR = "REGULAR"
    LARGE = "LARGE"

class TicketStatus(Enum):
    ACTIVE = "ACTIVE"
    PAID = "PAID"
    LOST = "LOST"

class PaymentMethod(Enum):
    CASH = "CASH"
    CREDIT_CARD = "CREDIT_CARD"
    DIGITAL_WALLET = "DIGITAL_WALLET"

# ============================================================
# Vehicle Class
# ============================================================

class Vehicle:
    def __init__(self, license_plate: str, vehicle_type: VehicleType, color: str):
        self.license_plate = license_plate
        self.vehicle_type = vehicle_type
        self.color = color

    def get_vehicle_type(self) -> VehicleType:
        return self.vehicle_type

    def get_license_plate(self) -> str:
        return self.license_plate

# ============================================================
# ParkingSpot Class
# ============================================================

class ParkingSpot:
    def __init__(self, spot_id: str, spot_type: SpotType, floor: int):
        self.spot_id = spot_id
        self.spot_type = spot_type
        self.floor = floor
        self.is_occupied = False
        self.vehicle: Optional[Vehicle] = None
        self.lock = Lock()  # Thread safety

    def can_fit_vehicle(self, vehicle_type: VehicleType) -> bool:
        """Check if vehicle can fit in this spot"""
        if vehicle_type == VehicleType.MOTORCYCLE:
            return True  # Can park anywhere
        elif vehicle_type == VehicleType.CAR:
            return self.spot_type in [SpotType.REGULAR, SpotType.LARGE]
        elif vehicle_type == VehicleType.TRUCK:
            return self.spot_type == SpotType.LARGE
        return False

    def assign_vehicle(self, vehicle: Vehicle) -> bool:
        """Thread-safe vehicle assignment"""
        with self.lock:
            if self.is_occupied:
                return False  # Already occupied
            
            if not self.can_fit_vehicle(vehicle.get_vehicle_type()):
                return False  # Vehicle doesn't fit
            
            self.vehicle = vehicle
            self.is_occupied = True
            return True

    def remove_vehicle(self) -> None:
        """Thread-safe vehicle removal"""
        with self.lock:
            self.vehicle = None
            self.is_occupied = False

    def get_spot_id(self) -> str:
        return self.spot_id

    def get_floor(self) -> int:
        return self.floor

    def get_spot_type(self) -> SpotType:
        return self.spot_type

# ============================================================
# ParkingTicket Class
# ============================================================

class ParkingTicket:
    def __init__(self, ticket_id: str, vehicle: Vehicle, spot: ParkingSpot):
        self.ticket_id = ticket_id
        self.vehicle = vehicle
        self.spot = spot
        self.entry_time = datetime.now()
        self.exit_time: Optional[datetime] = None
        self.status = TicketStatus.ACTIVE

    def calculate_duration(self) -> int:
        """Calculate parking duration in hours"""
        end_time = datetime.now() if self.status == TicketStatus.ACTIVE else self.exit_time
        duration = (end_time - self.entry_time).total_seconds() / 3600
        return max(1, int(duration))  # Minimum 1 hour

    def mark_exit(self) -> None:
        self.exit_time = datetime.now()
        self.status = TicketStatus.PAID

    def get_ticket_id(self) -> str:
        return self.ticket_id

    def get_vehicle(self) -> Vehicle:
        return self.vehicle

    def get_spot(self) -> ParkingSpot:
        return self.spot

    def get_status(self) -> TicketStatus:
        return self.status

# ============================================================
# Strategy Pattern: Pricing Strategy
# ============================================================

class PricingStrategy(ABC):
    @abstractmethod
    def calculate_fee(self, vehicle_type: VehicleType, hours: int) -> float:
        pass

class HourlyPricingStrategy(PricingStrategy):
    def calculate_fee(self, vehicle_type: VehicleType, hours: int) -> float:
        rates = {
            VehicleType.MOTORCYCLE: 2.0,
            VehicleType.CAR: 4.0,
            VehicleType.TRUCK: 6.0
        }
        hourly_rate = rates.get(vehicle_type, 4.0)
        return hourly_rate * hours

# ============================================================
# Payment Class
# ============================================================

class Payment:
    def __init__(self, amount: float, payment_method: PaymentMethod):
        self.amount = amount
        self.payment_method = payment_method
        self.timestamp = datetime.now()
        self.processed = False

    def process_payment(self) -> bool:
        """Simulate payment processing"""
        self.processed = True
        print(f"Payment of ${self.amount:.2f} processed successfully.")
        return True

    def get_amount(self) -> float:
        return self.amount

    def is_processed(self) -> bool:
        return self.processed

# ============================================================
# Observer Pattern: DisplayBoard
# ============================================================

class DisplayBoard:
    def __init__(self, floor_number: int):
        self.floor_number = floor_number
        self.available_spots: Dict[SpotType, int] = {
            SpotType.COMPACT: 0,
            SpotType.REGULAR: 0,
            SpotType.LARGE: 0
        }

    def update(self, spot_type: SpotType, count: int) -> None:
        self.available_spots[spot_type] = count
        self.display()

    def display(self) -> None:
        print(f"Floor {self.floor_number} - Available Spots: "
              f"Compact={self.available_spots[SpotType.COMPACT]}, "
              f"Regular={self.available_spots[SpotType.REGULAR]}, "
              f"Large={self.available_spots[SpotType.LARGE]}")

# ============================================================
# ParkingFloor Class
# ============================================================

class ParkingFloor:
    def __init__(self, floor_number: int):
        self.floor_number = floor_number
        self.spots: Dict[str, ParkingSpot] = {}
        self.display_board = DisplayBoard(floor_number)
        self.lock = RLock()

    def add_spot(self, spot: ParkingSpot) -> None:
        with self.lock:
            self.spots[spot.get_spot_id()] = spot
            self.update_display_board()

    def get_available_spot(self, vehicle_type: VehicleType) -> Optional[ParkingSpot]:
        """Find available spot for vehicle type"""
        with self.lock:
            for spot in self.spots.values():
                if not spot.is_occupied and spot.can_fit_vehicle(vehicle_type):
                    return spot
            return None

    def update_display_board(self) -> None:
        """Count available spots by type and update display"""
        counts = {
            SpotType.COMPACT: 0,
            SpotType.REGULAR: 0,
            SpotType.LARGE: 0
        }

        for spot in self.spots.values():
            if not spot.is_occupied:
                counts[spot.get_spot_type()] += 1

        # Update display board
        for spot_type, count in counts.items():
            self.display_board.update(spot_type, count)

    def get_floor_number(self) -> int:
        return self.floor_number

# ============================================================
# Singleton Pattern: ParkingLot
# ============================================================

class ParkingLot:
    _instance = None
    _lock = Lock()

    def __new__(cls, name: str = "Main Parking Lot", address: str = "123 Main St"):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super().__new__(cls)
                    cls._instance._initialized = False
        return cls._instance

    def __init__(self, name: str = "Main Parking Lot", address: str = "123 Main St"):
        if self._initialized:
            return
        
        self.name = name
        self.address = address
        self.floors: List[ParkingFloor] = []
        self.active_tickets: Dict[str, ParkingTicket] = {}
        self.pricing_strategy: PricingStrategy = HourlyPricingStrategy()
        self.ticket_counter = 0
        self.lot_lock = RLock()
        self._initialized = True

    @classmethod
    def get_instance(cls, name: str = "Main Parking Lot", address: str = "123 Main St"):
        """Thread-safe Singleton instance"""
        return cls(name, address)

    def add_floor(self, floor: ParkingFloor) -> None:
        with self.lot_lock:
            self.floors.append(floor)

    def get_available_spot(self, vehicle_type: VehicleType) -> Optional[ParkingSpot]:
        """Find available spot across all floors"""
        for floor in self.floors:
            spot = floor.get_available_spot(vehicle_type)
            if spot is not None:
                return spot
        return None

    def process_entry(self, vehicle: Vehicle) -> ParkingTicket:
        """Process vehicle entry"""
        with self.lot_lock:
            # Find available spot
            spot = self.get_available_spot(vehicle.get_vehicle_type())
            if spot is None:
                raise Exception("No available spot for vehicle type")

            # Assign vehicle to spot
            if not spot.assign_vehicle(vehicle):
                raise Exception("Failed to assign vehicle to spot")

            # Generate ticket
            self.ticket_counter += 1
            ticket_id = f"T{self.ticket_counter}"
            ticket = ParkingTicket(ticket_id, vehicle, spot)
            self.active_tickets[ticket_id] = ticket

            print(f"Vehicle {vehicle.get_license_plate()} "
                  f"parked at spot {spot.get_spot_id()} "
                  f"on floor {spot.get_floor()}")

            # Update display boards
            for floor in self.floors:
                floor.update_display_board()

            return ticket

    def process_exit(self, ticket: ParkingTicket) -> Payment:
        """Process vehicle exit"""
        with self.lot_lock:
            if ticket.get_status() != TicketStatus.ACTIVE:
                raise Exception("Ticket is not active")

            # Calculate fee
            hours = ticket.calculate_duration()
            fee = self.pricing_strategy.calculate_fee(
                ticket.get_vehicle().get_vehicle_type(), hours)

            # Create payment
            payment = Payment(fee, PaymentMethod.CREDIT_CARD)

            # Process payment
            if payment.process_payment():
                # Mark ticket as paid
                ticket.mark_exit()

                # Free the spot
                ticket.get_spot().remove_vehicle()

                # Remove from active tickets
                del self.active_tickets[ticket.get_ticket_id()]

                print(f"Vehicle {ticket.get_vehicle().get_license_plate()} "
                      f"exited. Duration: {hours} hours, Fee: ${fee:.2f}")

                # Update display boards
                for floor in self.floors:
                    floor.update_display_board()

            return payment

    def display_status(self) -> None:
        print("\n=== Parking Lot Status ===")
        print(f"Name: {self.name}")
        print(f"Active Tickets: {len(self.active_tickets)}")
        for floor in self.floors:
            floor.update_display_board()
        print("========================\n")

# ============================================================
# Main - Example Usage
# ============================================================

def main():
    try:
        # Get parking lot instance (Singleton)
        parking_lot = ParkingLot.get_instance("City Center Parking", "Downtown")

        # Create floors
        floor1 = ParkingFloor(1)
        floor2 = ParkingFloor(2)

        # Add spots to floor 1
        floor1.add_spot(ParkingSpot("F1-C1", SpotType.COMPACT, 1))
        floor1.add_spot(ParkingSpot("F1-C2", SpotType.COMPACT, 1))
        floor1.add_spot(ParkingSpot("F1-R1", SpotType.REGULAR, 1))
        floor1.add_spot(ParkingSpot("F1-R2", SpotType.REGULAR, 1))
        floor1.add_spot(ParkingSpot("F1-L1", SpotType.LARGE, 1))

        # Add spots to floor 2
        floor2.add_spot(ParkingSpot("F2-C1", SpotType.COMPACT, 2))
        floor2.add_spot(ParkingSpot("F2-R1", SpotType.REGULAR, 2))
        floor2.add_spot(ParkingSpot("F2-L1", SpotType.LARGE, 2))

        # Add floors to parking lot
        parking_lot.add_floor(floor1)
        parking_lot.add_floor(floor2)

        parking_lot.display_status()

        # Create vehicles
        car1 = Vehicle("ABC123", VehicleType.CAR, "Red")
        bike1 = Vehicle("XYZ789", VehicleType.MOTORCYCLE, "Black")
        truck1 = Vehicle("TRK456", VehicleType.TRUCK, "White")

        # Process entries
        ticket1 = parking_lot.process_entry(car1)
        ticket2 = parking_lot.process_entry(bike1)
        ticket3 = parking_lot.process_entry(truck1)

        parking_lot.display_status()

        # Simulate some time passing
        print("\n[Simulating time passage...]\n")
        time.sleep(1)

        # Process exits
        parking_lot.process_exit(ticket1)
        parking_lot.process_exit(ticket2)

        parking_lot.display_status()

        # Try to park another car
        car2 = Vehicle("DEF456", VehicleType.CAR, "Blue")
        ticket4 = parking_lot.process_entry(car2)

        parking_lot.display_status()

    except Exception as e:
        print(f"Error: {str(e)}")

if __name__ == "__main__":
    main()
